src/impl/tlstransport.cpp

```diff
@@ -296,7 +296,224 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+
+	PLOG_DEBUG << "Initializing TLS transport (MbedTLS)";
+
+	mbedtls_entropy_init(&mEntropy);
+	mbedtls_ctr_drbg_init(&mDrbg);
+	mbedtls_ssl_init(&mSsl);
+	mbedtls_ssl_config_init(&mConf);
+	mbedtls_ctr_drbg_set_prediction_resistance(&mDrbg, MBEDTLS_CTR_DRBG_PR_ON);
+
+	try {
+		mbedtls::check(mbedtls_ctr_drbg_seed(&mDrbg, mbedtls_entropy_func, &mEntropy, NULL, 0));
+
+		mbedtls::check(mbedtls_ssl_config_defaults(
+		    &mConf, mIsClient ? MBEDTLS_SSL_IS_CLIENT : MBEDTLS_SSL_IS_SERVER,
+		    MBEDTLS_SSL_TRANSPORT_STREAM, MBEDTLS_SSL_PRESET_DEFAULT));
+
+		mbedtls_ssl_conf_authmode(&mConf, MBEDTLS_SSL_VERIFY_OPTIONAL);
+		mbedtls_ssl_conf_rng(&mConf, mbedtls_ctr_drbg_random, &mDrbg);
+
+		if (certificate != nullptr) {
+			auto [crt, pk] = certificate->credentials();
+			mbedtls::check(mbedtls_ssl_conf_own_cert(&mConf, crt.get(), pk.get()));
+		}
+
+		mbedtls::check(mbedtls_ssl_setup(&mSsl, &mConf));
+		mbedtls_ssl_set_bio(&mSsl, static_cast<void *>(this), WriteCallback, ReadCallback, NULL);
+	} catch (...) {
+		mbedtls_entropy_free(&mEntropy);
+		mbedtls_ctr_drbg_free(&mDrbg);
+		mbedtls_ssl_free(&mSsl);
+		mbedtls_ssl_config_free(&mConf);
+		throw;
+	}
+}
+
+TlsTransport::~TlsTransport() {}
+
+void TlsTransport::start() {
+	PLOG_DEBUG << "Starting TLS transport";
+	registerIncoming();
+	changeState(State::Connecting);
+	enqueueRecv(); // to initiate the handshake
+}
+
+void TlsTransport::stop() {
+	PLOG_DEBUG << "Stopping TLS transport";
+	unregisterIncoming();
+	mIncomingQueue.stop();
+	enqueueRecv();
+}
+
+bool TlsTransport::send(message_ptr message) {
+	if (state() != State::Connected)
+		throw std::runtime_error("TLS is not open");
+
+	if (!message || message->size() == 0)
+		return outgoing(message); // pass through
+
+	PLOG_VERBOSE << "Send size=" << message->size();
+
+	mbedtls::check(mbedtls_ssl_write(
+	    &mSsl, reinterpret_cast<const unsigned char *>(message->data()), int(message->size())));
+
+	return true;
+}
+
+void TlsTransport::incoming(message_ptr message) {
+	if (!message) {
+		mIncomingQueue.stop();
+		enqueueRecv();
+		return;
+	}
+
+	PLOG_VERBOSE << "Incoming size=" << message->size();
+	mIncomingQueue.push(message);
+	enqueueRecv();
+}
+
+bool TlsTransport::outgoing(message_ptr message) { return Transport::outgoing(std::move(message)); }
+
+void TlsTransport::postHandshake() {
+	// Dummy
+}
+
+void TlsTransport::doRecv() {
+	std::lock_guard lock(mRecvMutex);
+	--mPendingRecvCount;
+
+	if (state() != State::Connecting && state() != State::Connected)
+		return;
+
+	try {
+		const size_t bufferSize = 4096;
+		char buffer[bufferSize];
+
+		// Handle handshake if connecting
+		if (state() == State::Connecting) {
+			auto handshakeStatus = mbedtls_ssl_handshake(&mSsl);
+			switch (handshakeStatus) {
+			case 0:
+				break;
+			case MBEDTLS_ERR_SSL_WANT_WRITE:
+			case MBEDTLS_ERR_SSL_WANT_READ:
+			case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+			case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS: {
```

Comment:
Could you please refactor this the same was as in `DtlsTransport`?

Reply:
Fixed!